datastore/longhorn.go

```diff
@@ -53,108 +52,138 @@ func (s *DataStore) UpdateCustomizedSettings(overrideDefaultSettings map[types.S
+				},
+				Value: definition.Default,
 			}
-		}
 
-		// Won't allow users tamper the read-only default settings
-		// TODO: Ideally, we should separate system settings and user settings.
-		if !definition.ReadOnly {
-			// If the longhorn-default-setting configmap is updated,
-			// then update the definition value and resource version.
-			if configMapResourceVersion != defaultSettingCM.ResourceVersion {
-				if value, exist := customizedDefaultSettings[string(sName)]; exist {
-					definition.Default = value
-					definition.ConfigMapResourceVersion = defaultSettingCM.ResourceVersion
-				}
+			if _, err := s.CreateSetting(setting); err != nil && !apierrors.IsAlreadyExists(err) {
+				return err
 			}
 		}
-
-		types.SetSettingDefinition(sName, definition)
 	}
 
 	return nil
 }
 
-func (s *DataStore) createOrUpdateSetting(name types.SettingName, definition types.SettingDefinition) error {
+func (s *DataStore) syncSettingsWithDefaultImages(defaultImage map[types.SettingName]string) error {
+	for name, value := range defaultImage {
+		if err := s.createOrUpdateSetting(name, value, ""); err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (s *DataStore) createOrUpdateSetting(name types.SettingName, value, defaultSettingCMResourceVersion string) error {
+	err := s.ValidateSetting(string(name), value)
+	if err != nil {
+		return err
+	}
+
 	setting, err := s.GetSettingExact(name)
 	if err != nil {
 		if !ErrorIsNotFound(err) {
 			return err
 		}
+
 		setting = &longhorn.Setting{
 			ObjectMeta: metav1.ObjectMeta{
 				Name:        string(name),
-				Annotations: map[string]string{types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey): definition.ConfigMapResourceVersion},
+				Annotations: map[string]string{types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey): defaultSettingCMResourceVersion},
 			},
-			Value: definition.Default,
+			Value: value,
 		}
 
-		if err := s.ValidateSetting(string(setting.Name), setting.Value); err != nil {
+		if _, err = s.CreateSetting(setting); err != nil && !apierrors.IsAlreadyExists(err) {
 			return err
 		}
+		return nil
+	}
 
-		_, err := s.CreateSetting(setting)
-		if err != nil && !apierrors.IsAlreadyExists(err) {
-			return err
+	if setting.Annotations == nil {
+		setting.Annotations = map[string]string{}
+	}
+
+	setting.Annotations[types.GetLonghornLabelKey(types.ConfigMapResourceVersionKey)] = defaultSettingCMResourceVersion
+	setting.Value = value
+	_, err = s.UpdateSetting(setting)
+	return err
+}
+
+func (s *DataStore) applyCustomizedDefaultSettingsToDefinitions(customizedDefaultSettings map[string]string) error {
+	for _, sName := range types.SettingNameList {
+		definition, ok := types.GetSettingDefinition(sName)
+		if !ok {
+			return fmt.Errorf("BUG: setting %v is not defined", sName)
 		}
 
-		return nil
+		if value, ok := customizedDefaultSettings[string(sName)]; ok {
+			definition.Default = value
+			types.SetSettingDefinition(sName, definition)
+		}
```

Comment:
Maybe it's better to check if the field is read-only before applying the customization from ConfigMap.

Reply:
Good idea. Thank you.